### PR TITLE
[Runtime] Use GIL lock in the runtime instead of custom mutexes.

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** We have a custom mutex that denotes current use of the Python interpreter (a shared resource among threads). This custom mutex may be replaced by the lock found in pybind11.

**Description of the Change:** Use the scope acquire lock.

**Benefits:** No custom mutex, better DX with third party libraries. (No need to determine location of the mutex library).

**Possible Drawbacks:**

**Related GitHub Issues:**

TODO:

- [ ] ~~add a python file and outline all strings that are input to exec~~
- [x] OQC device
- [x] release gil when executing the memory transfer
- [x] write functions and classes for dlopen dlclose

### CODE REVIEW

**Comment by maliasadi** [comment]
:raised_hands:

**Comment by erick-xanadu** [comment]
@maliasadi still under construction :sweat_smile:

**Comment by maliasadi** [comment]
I know.. I just got too excited :sweat_smile:

**Comment by dime10** [comment]
@erick-xanadu This seemed like a good thing, any chance we can still land this?

**Comment by erick-xanadu** [comment]
> @erick-xanadu This seemed like a good thing, any chance we can still land this?

@dime10, needs some more investigation.

**Comment by erick-xanadu** [comment]
@joeycarter thanks! I added you as a reviewer because you will be working on something similar this quarter. We don't have to merge this PR as is, but I think this essentially decouples the runtime from python as much as possible while maintaining the current functionality. Regarding dlopen + ASan, I am not 100% sure yet why that is the case. We can investigate in a few days :)

**Comment by dime10** [comment]
Please let me know if you plan to merge this at some point, I'd like to take a look first.

**Comment by erick-xanadu** [comment]
> Please let me know if you plan to merge this at some point, I'd like to take a look first.

I'll leave it up to @joeycarter and you to decide if you want to merge this. This implements the decoupling between the Python API and our runtime without removing any functionality. (Modulo TODO OQD). This means that our runtime can be compiled without linking against pybind11 and the Python C-API. All of this is achieved by:

1. Moving everything that was pybind into its own module.
2. Having functions with default visibility in the pybind module.
3. Building the pybind module with pybind.
4. Building the runtime without pybind.
5. `dlopen`ing the pybind module when executing the OpenQASM support.
6. `dlsym`ing the functions we want to call from the pybind module.
7. We ensure that when we enter the functions with default visibility we acquire the GIL from python.

What I would be nice at some point is to move the python program which is embedded as a string in the C++ pybind files to a python file. This would allow me to just have a pybind file that `import`s that python module and call the functions. But I found that to be more hassle than it is worth. By keeping the pybind module a C++ file, I am able to use the `$ORIGIN` rpath value to find the pybind module at runtime and `dlopen` it. But python modules do not use `$ORIGIN`, instead it uses `$PYTHONPATH`. But I cannot set the `$PYTHONPATH` variable in a nice manner that plays nicely with the build system.

**Comment by erick-xanadu** [comment]
@joeycarter , there are a couple of minor TODOs. One is that instead of using dlopen / dlclose (which I think I forgot to use dlclose...) is to use a C++ struct whose constructor uses dlopen and destructor uses dlclose. The other one, is about the OQC file that lives in `frontend/catalyst/third_party/oqc/src/OQCRunner.hpp`. We just need to convert that to something similar to what was done for the OpenQASM device.

File: frontend/catalyst/third_party/oqc/src/CMakeLists.txt [hunk 1]
```diff
@@ -2,14 +2,10 @@ cmake_minimum_required(VERSION 3.20)
 
 project(catalyst_oqc)
 
-option(USE_ALTERNATIVE_CATALYST_PYTHON_INTERPRETER_PATH OFF)
```
[the remark this answers is not in the document]
Comment (joeycarter) [reply frontend/catalyst/third_party/oqc/src/CMakeLists.txt:5]: Fixed in 2a3459d.

**Comment by joeycarter** [comment]
Hey @erick-xanadu, I've completed the two TODOs, could you take a look when you get the chance?

**Comment by erick-xanadu** [comment]
@joeycarter , it looks like some difference in macOS and ubuntu... I'll investigate a bit, but if I don't solve it we should just reintroduce the lock if that will solve the issue. The story assigned this quarter is about removing pybind and not the lock.

File: runtime/include/DynamicLibraryLoader.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        if (!handle) {
+            const char *err_msg = dlerror();
+            RT_FAIL(err_msg);
+        }
+    }
+
+    ~DynamicLibraryLoader()
+    {
+        if (handle) {
+            dlclose(handle);
```
Comment (erick-xanadu) [inline runtime/include/DynamicLibraryLoader.hpp:42]:
```suggestion
            // dlclose(handle); TODO: investigate
```

File: .github/workflows/check-catalyst.yaml [hunk 3]
```diff
@@ -647,10 +647,11 @@ jobs:
       - name: Build Runtime test suite for OpenQasm device
         if: ${{ matrix.backend == 'openqasm' }}
         run: |
+          # Asan prevents dlopen from working?
           C_COMPILER=$(which ${{ needs.constants.outputs[format('c_compiler.{0}', matrix.compiler)] }}) \
           CXX_COMPILER=$(which ${{ needs.constants.outputs[format('cxx_compiler.{0}', matrix.compiler)] }}) \
           COMPILER_LAUNCHER="" \
-          ENABLE_ASAN=ON \
+          ENABLE_ASAN=OFF \
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply .github/workflows/check-catalyst.yaml:645]: Yep

**Comment by erick-xanadu** [comment]
@joeycarter I'll let you merge this. This looks good on my side, even though dlclose is not called. We can also make a macro to limit this behaviour only on macOS. Your call :)

**Comment by joeycarter** [comment]
Hi @erick-xanadu, thanks for the investigative work! I opted to wrap `dlcose` in the macro:

```cpp
if (handle) {
    ...
#ifndef __APPLE__
    dlclose(handle);
#endif
}
```

At least we can do it "properly" on Linux :slightly_smiling_face:

I'm happy to merge this is :100:

**Pull request merged**